### DISCUSSION TITLE
Identifying Summary Nodes

### DISCUSSION THREAD

**Discussion by timboxyz** [freeplane/freeplane#586] [Help]
I have a script that traverses back from a leaf node to the root, extracting route information on the way.

My maps often contain summary nodes where, having expanded details, I collapse them again to a single line. The script happily works back past any summary nodes, but I need to extract some of the node detail of summarised nodes. This would be easy to do if I could spot a summary node, and then list its siblings.

AFAIK a summary node is in fact 2 nodes either side of a group of siblings with the visible summary node attached to one of them. I am trying to find an attribute that enables me to spot when I have hit a summary node so I can then peek at its siblings.

Can anyone suggest some identifying feature that will allow me to do this?

**Comment by macmarrum** [comment]
> AFAIK a summary node is in fact 2 nodes either side of a group of siblings with the visible summary node attached to one of them. 

Yep. When I look at the xml of my simple .mm file containing a summary node, this is what I can see
```xml
<node TEXT="A" POSITION="right" ID="ID_1319294995"/>
<node TEXT="" POSITION="right" ID="ID_1439033905"><!-- a technical node to start a summary group -->
    <hook NAME="FirstGroupNode"/>
</node>
<node TEXT="B" POSITION="right" ID="ID_1150398632"/>
<node TEXT="C" POSITION="right" ID="ID_1411160372"/>
<node TEXT="" POSITION="right" ID="ID_1136607364"><!-- a technical node to end a summary group -->
    <hook NAME="SummaryNode"/>
    <hook NAME="AlwaysUnfoldedNode"/>
    <node TEXT="S" ID="ID_1013360672"/><!-- a node spawning from the brace tip -->
</node>
<node TEXT="D" POSITION="right" ID="ID_1671085766"/>
```


Scripting API is missing information about summary nodes, except that they aren't visible.
If you're feeling adventurous, you can use Freeplane's internal API. The following static methods are available
* `org.freeplane.features.map.SummaryNode#isSummaryNode`
* `org.freeplane.features.map.SummaryNode#isFirstGroupNode`

```groovy
import static org.freeplane.features.map.SummaryNode.isSummaryNode

node.pathToRoot.reverse().each { // pathToRoot is actually path from root to the node, hence reverse()
    if (isSummaryNode(it.delegate)) {  // isSummaryNode takes NodeModel, hence node.delegate
        it // a technical node
        it.children[0] // the first node spawning from the brace tip
    }
}
```

Just keep in mind that the internal API can change with a new FP version, without notice, and it's your responsibility to update your scripts, if that happens. This is why I said _if you are adventurous_ :wink: 
But in reality, such changes are quite rare.

**Reply by timboxyz** [reply]
Thanks, that's very helpful, as you say the API lacks information so good to know there is stuff under the hood to get at.
